File: lib/web/browser_access.rb
require "json"
require "securerandom"
require_relative "../browser_access_store"

module Web
  module BrowserAccess
    ACCESS_ENDPOINTS = %w[
      /browser-access/request
      /browser-access/admin-login
      /browser-access/status
      /browser-access/pending
      /browser-access/approve
      /browser-access/deny
    ].freeze
    BROWSER_ACCESS_STORE_CACHE = {}
    BROWSER_ACCESS_STORE_CACHE_MUTEX = Mutex.new

    module Helpers
      private

      def browser_access_enabled?
        !settings.browser_auth_disabled && !settings.gateway_admin_password.to_s.empty?
      end

      def browser_access_store
        path = settings.browser_access_path
        BrowserAccess::BROWSER_ACCESS_STORE_CACHE_MUTEX.synchronize do
          BrowserAccess::BROWSER_ACCESS_STORE_CACHE[path] ||= BrowserAccessStore.new(path: path)
        end
      end

      def browser_token
        return @browser_token if defined?(@browser_token)

        @browser_token = submitted_browser_token
        return @browser_token if @browser_token

        set_browser_token(SecureRandom.hex(32))
      end

      def submitted_browser_token
        token = request.cookies["gripi_browser"].to_s
        token unless token.empty? || token.bytesize > 128
      end

      def set_browser_token(token)
        @browser_token = token
        response.set_cookie("gripi_browser", value: token, path: "/", httponly: true, secure: secure_transport?, same_site: :lax, max_age: 365 * 24 * 60 * 60)
        token
      end

      def approved_browser?
        browser_access_enabled? && browser_access_store.approved?(browser_token)
      end

      def enforce_browser_access
        return unless browser_access_enabled?
        return if multi_user_mode?
        return if BrowserAccess::ACCESS_ENDPOINTS.include?(request.path_info)
        return if approved_browser?

        @access_request = browser_access_store.pending_request(browser_token)
        @access_return_to = request.fullpath
        @access_error = nil
        status 403
        halt erb(:access_blocked)
      end

      def safe_return_to
        return_to = params["return_to"].to_s
        return return_to if return_to.bytesize <= 2_048 && return_to.start_with?("/") && !return_to.start_with?("//")

        "/"
      end

      def client_rate_limit_key
        request.env["REMOTE_ADDR"].to_s
      end

      def valid_access_code?(code)
        code.to_s.match?(/\A[A-Z0-9]{4}-[A-Z0-9]{4}\z/)
      end

      def secure_compare(left, right)
        return false if left.empty? || right.empty?
        return false unless left.bytesize == right.bytesize

        Rack::Utils.secure_compare(left, right)
      end
    end

    def self.registered(app)
      app.helpers Helpers

      app.post "/browser-access/request" do
        halt 404 unless browser_access_enabled?

        halt 429, "Too many access requests" unless settings.access_request_rate_limiter.allow?(client_rate_limit_key)
        begin
          browser_access_store.request_access(browser_token, ip: request.ip, user_agent: request.user_agent)
        rescue BrowserAccessStore::PendingRequestsFull
          @access_request = nil
          @access_return_to = safe_return_to
          @access_error = "Too many pending browser access requests. Try again after an existing request is approved or denied."
          status 503
          halt erb(:access_blocked)
        end
        redirect safe_return_to
      end

      app.post "/browser-access/admin-login" do
        halt 404 unless browser_access_enabled?

        halt 429, "Too many admin login attempts" unless settings.admin_login_rate_limiter.allow?(client_rate_limit_key)

        if secure_compare(params["password"].to_s, settings.gateway_admin_password.to_s)
          old_token = submitted_browser_token
          new_token = SecureRandom.hex(32)
          browser_access_store.replace_browser_token(old_token, new_token, label: request.user_agent)
          set_browser_token(new_token)
          redirect safe_return_to
        else
          @access_request = browser_access_store.pending_request(browser_token)
          @access_return_to = safe_return_to
          @access_error = "Admin password did not match."
          status 403
          erb :access_blocked
        end
      end

      app.get "/browser-access/status" do
        halt 404 unless browser_access_enabled?

        content_type :json
        JSON.generate(status: browser_access_store.pending_status(browser_token))
      end

      app.get "/browser-access/pending" do
        halt 403 unless approved_browser?

        content_type :json
        JSON.generate(requests: browser_access_store.pending_requests)
      end

      app.post "/browser-access/approve" do
        halt 403 unless approved_browser?

        halt 400, "Valid code is required" unless valid_access_code?(params["code"])

        request = browser_access_store.approve_code(params.fetch("code"))
        content_type :json
        JSON.generate(ok: !request.nil?)
      end

      app.post "/browser-access/deny" do
        halt 403 unless approved_browser?

        halt 400, "Valid code is required" unless valid_access_code?(params["code"])

        request = browser_access_store.deny_code(params.fetch("code"))
        content_type :json
        JSON.generate(ok: !request.nil?)
      end
    end
  end
end

File: internal/server/browser_access.go
package server

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"net/netip"
	"regexp"
	"strings"

	"github.com/melounvitek/gripi/internal/access"
)

var accessCodePattern = regexp.MustCompile(`^[A-Z0-9]{4}-[A-Z0-9]{4}$`)

var browserAccessPaths = map[string]struct{}{
	"/browser-access/request":     {},
	"/browser-access/admin-login": {},
	"/browser-access/status":      {},
	"/browser-access/pending":     {},
	"/browser-access/approve":     {},
	"/browser-access/deny":        {},
}

type accessPageData struct {
	Request           *access.PendingRequest
	ReturnTo          string
	Error             string
	BrowserAccessPath string
	Nonce             string
}

func (app *application) registerBrowserAccessRoutes(mux *http.ServeMux) {
	mux.HandleFunc("POST /browser-access/request", app.requestBrowserAccess)
	mux.HandleFunc("/browser-access/request", http.NotFound)
	mux.HandleFunc("POST /browser-access/admin-login", app.adminLogin)
	mux.HandleFunc("/browser-access/admin-login", http.NotFound)
	mux.HandleFunc("GET /browser-access/status", app.browserAccessStatus)
	mux.HandleFunc("/browser-access/status", http.NotFound)
	mux.HandleFunc("GET /browser-access/pending", app.pendingBrowserAccess)
	mux.HandleFunc("/browser-access/pending", http.NotFound)
	mux.HandleFunc("POST /browser-access/approve", app.approveBrowserAccess)
	mux.HandleFunc("/browser-access/approve", http.NotFound)
	mux.HandleFunc("POST /browser-access/deny", app.denyBrowserAccess)
	mux.HandleFunc("/browser-access/deny", http.NotFound)
}

func (app *application) enforceBrowserAccess(next http.Handler) http.Handler {
	if !app.browserAccessEnabled() || app.config.MultiUserMode {
		return next
	}
	return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
		if _, accessEndpoint := browserAccessPaths[request.URL.Path]; accessEndpoint || staticAssetPath(request.URL.Path) {
			next.ServeHTTP(response, request)
			return
		}
		token, err := app.browserToken(response, request)
		if err != nil {
			writeText(response, http.StatusInternalServerError, "Internal Server Error")
			return
		}
		approved, err := app.browserStore.Approved(token)
		if err != nil {
			writeText(response, http.StatusInternalServerError, "Internal Server Error")
			return
		}
		if approved {
			next.ServeHTTP(response, request)
			return
		}
		pending, found, err := app.browserStore.PendingRequest(token)
		if err != nil {
			writeText(response, http.StatusInternalServerError, "Internal Server Error")
			return
		}
		var pendingPointer *access.PendingRequest
		if found {
			pendingPointer = &pending
		}
		returnTo := request.URL.RequestURI()
		if returnTo == "" {
			returnTo = "/"
		}
		app.renderAccessPage(response, request, http.StatusForbidden, pendingPointer, returnTo, "")
	})
}

func (app *application) requestBrowserAccess(response http.ResponseWriter, request *http.Request) {
	if !app.browserAccessEnabled() {
		http.NotFound(response, request)
		return
	}
	if !app.accessLimiter.Allow(clientIP(request.RemoteAddr)) {
		writeText(response, http.StatusTooManyRequests, "Too many access requests")
		return
	}
	if !parseForm(response, request) {
		return
	}
	token, err := app.browserToken(response, request)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	_, err = app.browserStore.RequestAccess(token, app.browserRequestIP(request), request.UserAgent())
	if err != nil {
		var full *access.PendingRequestsFullError
		if errors.As(err, &full) {
			app.renderAccessPage(
				response,
				request,
				http.StatusServiceUnavailable,
				nil,
				safeReturnTo(request.Form.Get("return_to")),
				"Too many pending browser access requests. Try again after an existing request is approved or denied.",
			)
			return
		}
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	redirect(response, request, safeReturnTo(request.Form.Get("return_to")), app.config.TrustProxyHeaders)
}

func (app *application) adminLogin(response http.ResponseWriter, request *http.Request) {
	if !app.browserAccessEnabled() {
		http.NotFound(response, request)
		return
	}
	if !app.adminLimiter.Allow(clientIP(request.RemoteAddr)) {
		writeText(response, http.StatusTooManyRequests, "Too many admin login attempts")
		return
	}
	if !parseForm(response, request) {
		return
	}
	returnTo := safeReturnTo(request.Form.Get("return_to"))
	if secureEqual(request.Form.Get("password"), app.config.AdminPassword) {
		oldToken, _ := submittedBrowserToken(request)
		newToken, err := app.newBrowserToken()
		if err != nil {
			writeText(response, http.StatusInternalServerError, "Internal Server Error")
			return
		}
		if _, err := app.browserStore.ReplaceBrowserToken(oldToken, newToken, request.UserAgent()); err != nil {
			writeText(response, http.StatusInternalServerError, "Internal Server Error")
			return
		}
		app.setBrowserToken(response, request, newToken)
		redirect(response, request, returnTo, app.config.TrustProxyHeaders)
		return
	}

	token, err := app.browserToken(response, request)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	pending, found, err := app.browserStore.PendingRequest(token)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	var pendingPointer *access.PendingRequest
	if found {
		pendingPointer = &pending
	}
	app.renderAccessPage(response, request, http.StatusForbidden, pendingPointer, returnTo, "Admin password did not match.")
}

func (app *application) browserAccessStatus(response http.ResponseWriter, request *http.Request) {
	if !app.browserAccessEnabled() {
		http.NotFound(response, request)
		return
	}
	token, err := app.browserToken(response, request)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	status, err := app.browserStore.PendingStatus(token)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	writeJSON(response, struct {
		Status access.Status `json:"status"`
	}{status})
}

func (app *application) pendingBrowserAccess(response http.ResponseWriter, request *http.Request) {
	if !app.requireApprovedBrowser(response, request) {
		return
	}
	requests, err := app.browserStore.PendingRequests()
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	writeJSON(response, struct {
		Requests []access.PendingRequest `json:"requests"`
	}{requests})
}

func (app *application) approveBrowserAccess(response http.ResponseWriter, request *http.Request) {
	app.resolveBrowserAccess(response, request, app.browserStore.ApproveCode)
}

func (app *application) denyBrowserAccess(response http.ResponseWriter, request *http.Request) {
	app.resolveBrowserAccess(response, request, app.browserStore.DenyCode)
}

func (app *application) resolveBrowserAccess(response http.ResponseWriter, request *http.Request, resolve func(string) (access.PendingRequest, bool, error)) {
	if !app.requireApprovedBrowser(response, request) {
		return
	}
	if !parseForm(response, request) {
		return
	}
	code := request.Form.Get("code")
	if !accessCodePattern.MatchString(code) {
		writeText(response, http.StatusBadRequest, "Valid code is required")
		return
	}
	_, found, err := resolve(code)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	writeJSON(response, struct {
		OK bool `json:"ok"`
	}{found})
}

func (app *application) requireApprovedBrowser(response http.ResponseWriter, request *http.Request) bool {
	if !app.browserAccessEnabled() {
		writeText(response, http.StatusForbidden, "Forbidden")
		return false
	}
	approved, err := app.approvedBrowser(response, request)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return false
	}
	if !approved {
		writeText(response, http.StatusForbidden, "Forbidden")
		return false
	}
	return true
}

func (app *application) approvedBrowser(response http.ResponseWriter, request *http.Request) (bool, error) {
	token, err := app.browserToken(response, request)
	if err != nil {
		return false, err
	}
	return app.browserStore.Approved(token)
}

func (app *application) browserAccessEnabled() bool {
	return !app.config.BrowserAuthDisabled && app.config.AdminPassword != ""
}

func (app *application) browserRequestIP(request *http.Request) string {
	if app.config.TrustProxyHeaders {
		forwarded := firstForwardedValue(request.Header.Get("X-Forwarded-For"))
		if address, err := netip.ParseAddr(forwarded); err == nil {
			return address.String()
		}
	}
	return clientIP(request.RemoteAddr)
}

func (app *application) browserToken(response http.ResponseWriter, request *http.Request) (string, error) {
	if token, valid := submittedBrowserToken(request); valid {
		return token, nil
	}
	token, err := app.newBrowserToken()
	if err != nil {
		return "", err
	}
	app.setBrowserToken(response, request, token)
	return token, nil
}

func submittedBrowserToken(request *http.Request) (string, bool) {
	cookie, err := request.Cookie("gripi_browser")
	if err != nil || cookie.Value == "" || len(cookie.Value) > 128 {
		return "", false
	}
	return cookie.Value, true
}

func (app *application) setBrowserToken(response http.ResponseWriter, request *http.Request, token string) {
	http.SetCookie(response, &http.Cookie{
		Name:     "gripi_browser",
		Value:    token,
		Path:     "/",
		MaxAge:   365 * 24 * 60 * 60,
		HttpOnly: true,
		Secure:   app.secureTransport(request),
		SameSite: http.SameSiteLaxMode,
	})
}

func (app *application) renderAccessPage(response http.ResponseWriter, request *http.Request, status int, pending *access.PendingRequest, returnTo, message string) {
	response.Header().Set("Content-Type", "text/html; charset=utf-8")
	response.WriteHeader(status)
	_ = app.templates.ExecuteTemplate(response, "access_blocked.html", accessPageData{
		Request:           pending,
		ReturnTo:          returnTo,
		Error:             message,
		BrowserAccessPath: app.config.BrowserAccessPath,
		Nonce:             requestNonce(request),
	})
}

func safeReturnTo(value string) string {
	if len(value) <= 2048 && strings.HasPrefix(value, "/") && !strings.HasPrefix(value, "//") {
		return value
	}
	return "/"
}

func secureEqual(left, right string) bool {
	leftDigest := sha256.Sum256([]byte(left))
	rightDigest := sha256.Sum256([]byte(right))
	return subtle.ConstantTimeCompare(leftDigest[:], rightDigest[:]) == 1
}

func randomBrowserToken() (string, error) {
	value := make([]byte, 32)
	if _, err := rand.Read(value); err != nil {
		return "", err
	}
	return hex.EncodeToString(value), nil
}

func redirect(response http.ResponseWriter, request *http.Request, target string, trustProxy bool) {
	response.Header().Set("Content-Type", "text/html;charset=utf-8")
	response.Header().Set("Location", absoluteRedirectURL(request, target, trustProxy))
	response.WriteHeader(http.StatusSeeOther)
}

func writeJSON(response http.ResponseWriter, value any) {
	contents, err := json.Marshal(value)
	if err != nil {
		writeText(response, http.StatusInternalServerError, "Internal Server Error")
		return
	}
	if response.Header().Get("Content-Type") == "" {
		response.Header().Set("Content-Type", "application/json")
	}
	_, _ = response.Write(contents)
}

func staticAssetPath(path string) bool {
	if strings.HasPrefix(path, "/assets/") {
		return true
	}
	switch path {
	case "/apple-touch-icon.png", "/manifest.webmanifest", "/app-icon.svg", "/app-icon-maskable.svg", "/service-worker.js":
		return true
	default:
		return false
	}
}
